### PR TITLE
LF: fix complexity of TransactionBuilder#addNode

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
Fix issue with slowSnoc underlines by Martin in #7835

CHANGELOG_BEGIN
CHANGELOG_END

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/transaction-test-lib/src/main/scala/lf/transaction/test/TransactionBuilder.scala [hunk 1]
```diff
@@ -34,42 +36,38 @@ final class TransactionBuilder(pkgTxVersion: Ref.PackageId => TransactionVersion
...
-    val builtRoots = roots.result()
+    val builtNodes = nodes.transform {
+      case (nid, VersionedNode(version, exe: TxExercise)) =>
+        VersionedNode[NodeId, ContractId](version, exe.copy(children = children(nid).toImmArray))
+      case (_, node) =>
+        node
+    }
     val nodesVersions =
-      builtRoots.reverseIterator
+      roots.reverseIterator
```
Comment (hurryabit) [inline daml-lf/transaction-test-lib/src/main/scala/lf/transaction/test/TransactionBuilder.scala:64]: I imagine iterating over a `BackStack` from either side to be more expensive than iterating over an `ImmArray`. If that's the case, we should convert to an `ImmArray`, which we are doing anyway, first and then iterate over that.

File: daml-lf/transaction-test-lib/src/main/scala/lf/transaction/test/TransactionBuilder.scala [hunk 2]
```diff
@@ -34,42 +36,38 @@ final class TransactionBuilder(pkgTxVersion: Ref.PackageId => TransactionVersion
...
   }
 
   def build(): Tx.Transaction = ids.synchronized {
     import VersionTimeline.Implicits._
-    val builtNodes = nodes.result()
-    val builtRoots = roots.result()
+    val builtNodes = nodes.transform {
+      case (nid, VersionedNode(version, exe: TxExercise)) =>
+        VersionedNode[NodeId, ContractId](version, exe.copy(children = children(nid).toImmArray))
+      case (_, node) =>
```
Comment (hurryabit) [inline daml-lf/transaction-test-lib/src/main/scala/lf/transaction/test/TransactionBuilder.scala:60]: I find this somewhat dangerous but given that Scala uses linear search for pattern matching, we probably still need to do this for the sake of performance. 😿

**Review by hurryabit — APPROVED** [review]
This looks very reasonable to me. Thanks.

File: daml-lf/transaction-test-lib/src/main/scala/lf/transaction/test/TransactionBuilder.scala (hunk 1 above)
Comment (remyhaemmerle-da) [reply daml-lf/transaction-test-lib/src/main/scala/lf/transaction/test/TransactionBuilder.scala]: `Backstack` has only `reverseIterator`. I think this is pretty efficient. 
Anyway, as we convert roots to an `ImmArray`, just below I applied your suggestion.

File: daml-lf/transaction-test-lib/src/main/scala/lf/transaction/test/TransactionBuilder.scala (hunk 2 above)
Comment (remyhaemmerle-da) [reply daml-lf/transaction-test-lib/src/main/scala/lf/transaction/test/TransactionBuilder.scala]: Is it better like that ? 

```scala
  val finalNodes = nodes.transform {
      case (nid, VersionedNode(version, exe: TxExercise)) =>
        VersionedNode[NodeId, ContractId](version, exe.copy(children = children(nid).toImmArray))
      case (_, node @ VersionedNode(_, _: Node.LeafOnlyNode[_, _])) =>
        node
    }
```

Comment (hurryabit) [reply daml-lf/transaction-test-lib/src/main/scala/lf/transaction/test/TransactionBuilder.scala]: Yes, that looks better.

**Pull request merged**